lib/lambda_ethereum_consensus/state_transition/cache.ex

```diff
@@ -0,0 +1,91 @@
+defmodule LambdaEthereumConsensus.StateTransition.Cache do
+  @moduledoc """
+  Caches expensive function calls.
+  """
+
+  alias LambdaEthereumConsensus.StateTransition.{Accessors, Misc}
+
+  @spec init_cache_tables() :: :ok
+  def init_cache_tables do
+    [:total_active_balance, :beacon_proposer_index, :beacon_committee, :active_validator_count]
+    |> Enum.each(fn table ->
+      :ets.new(table, [:set, :public, :named_table])
+    end)
+  end
+
+  defp lookup(table, key, compute_fun) do
+    case :ets.lookup(table, key) do
+      [{^key, balance}] ->
+        balance
+
+      [] ->
+        balance = compute_fun.()
+        :ets.insert_new(table, {key, balance})
+        balance
+    end
+  end
+
+  @spec cache_total_active_balance(SszTypes.BeaconState.t(), (-> SszTypes.gwei())) ::
+          SszTypes.gwei()
+  def cache_total_active_balance(state, compute_fun) do
+    epoch = Accessors.get_current_epoch(state)
+    root = get_epoch_root(state, epoch)
+
+    lookup(:total_active_balance, {epoch, root}, compute_fun)
+  rescue
```

Comment:
can you achieve the same using :ok, :error? With this, you might be swallowing other errors

Reply:
Done